functions/geometry_vi.zsh

```diff
@@ -0,0 +1,25 @@
+# A geometry plugin to display the ZLE vi-mode state.
+
+update_rprompt() {
+    RPROMPT=${${RPROMPT/\[NORMAL\]/}/\[INSERT\]/}
+    RPROMPT="[$1]$RPROMPT"
+}
+
+update_prompt() {
+    PROMPT=${${PROMPT/\[NORMAL\]/}/\[INSERT\]/}
+    PROMPT="[$1]$PROMPT"
+}
+
+function zle-keymap-select {
+    keymap=$([[ $KEYMAP == vicmd ]] && echo "NORMAL" || echo "INSERT")
+    (( ${GEOMETRY_RPROMPT[(i)geometry_vi]} <= ${#GEOMETRY_RPROMPT} )) && update_rprompt $keymap
+    (( ${GEOMETRY_PROMPT[(i)geometry_vi]} <= ${#GEOMETRY_PROMPT} )) && update_prompt $keymap
+    zle reset-prompt
+}
+
+zle -N zle-keymap-select
```

Comment:
I'd actually like this to remove the keymap on load, and then only enable if the render function is called in a prompt.

Reply:
I think this would be problematic for people who already have another handler for keymap-select, but it made me think that we probably do need more logic to check and possibly chain the existing handler. For instance, I use `zsh_highlight` and I already have a zle-keymap-select handler which gets overridden:

```
$  zle -l | grep keymap-select
orig-s0.0000030000-r24711-zle-keymap-select (zle-keymap-select)
zle-keymap-select (_zsh_highlight_widget_orig-s0.0000030000-r24711-zle-keymap-select)

```

Reply:
Additionally, while working on moving the logic into the render call, I was thinking that `geometry_vi()` will be doing useless work if we keep checking/enabling the keymap-select widget on every render tick. Ideally, plugins that are not in any PROMPT array should not even get sourced? (I just checked the code an it looks like the `geometry.zsh` is sourcing `functions/*.zsh` right now, but this could be changed unless there's a fundamental reason to do it that way? In fact, if we only source used functions, it lets function authors have one-shot initialization logic in the function definition itself.

Let me know what you think.
Alex

Reply:
Definitely like your reasoning here. I moved from sourcing to autoload, now I think we should chain when adding/removing hooks to preserve the user environment.

Reply:
I saw that you added the `autoload` support. I wasn't ready to update my PR yet because I hadn't really tested with it, but now I'll be able to rebase this and add the hook chaining logic to get a working  vi mode :)

I'll try to update the PR before the end of the week, but it will depend how quickly I can clear my backlog.

Cheers,